test: replace `strictEqual` with `deepStrictEqual` for better assertion accuracy

lumirlumir opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Update or create tests
    - If performance-related, include a benchmark
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What is the purpose of this pull request?

Hello,

In this PR, I've updated several test assertions in `tests/languages/json-source-code.test.js` to use `assert.deepStrictEqual` instead of `assert.strictEqual`. 

This change ensures that complex objects are compared deeply rather than by reference, improving the accuracy of the tests.

#### What changes did you make? (Give an overview)

#### Related Issues

<!-- include tags like "fixes #123" or "refs #123" -->

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

mdjermanovic commented on tests/languages/json-source-code.test.js:
```diff
@@ -34,7 +34,7 @@ describe("JSONSourceCode", () => {
 
 			assert.strictEqual(sourceCode.constructor.name, "JSONSourceCode");
-			assert.strictEqual(sourceCode.ast, ast);
+			assert.deepStrictEqual(sourceCode.ast, ast);
```
Isn't `strictEqual` more accurate, as it ensures that `sourceCode.ast === ast`?

mdjermanovic replied:
In other words, if two objects are `strictEqual`, then they must be `deepStrictEqual` because they are the same object. The opposite doesn't apply, so `strictEqual` checks more than `deepStrictEqual`.

mdjermanovic replied:
Can you provide an example where replacing `strictEqual(obj1, obj2)` with `deepStrictEqual(obj1, obj2)` catches a difference that wouldn't be caught by `strictEqual(obj1, obj2)`?

lumirlumir replied:
Sorry, I implemented this test case incorrectly. What I actually wanted to check was something like this:

```
assert.deepStrictEqual(sourceCode.ast, {
	type: "Document",
	body: {
		type: "Object",
		properties: [],
	},
	tokens: [],
})
```

> In other words, if two objects are strictEqual, then they must be deepStrictEqual because they are the same object. The opposite doesn't apply (two objects can be deepStrictEqual but not strictEqual), so strictEqual checks more than deepStrictEqual.

You're right about that.

What I wanted to verify is that the AST remains unchanged after going through the JSONSourceCode class.

lumirlumir commented:
@mdjermanovic Thanks for your guidance — it gave me a good opportunity to think more carefully about deep vs. shallow equality.

I've added a commit that addresses your comment, and this change doesn't affect the existing behavior.
